### ISSUE TITLE
Strange spack/build_environment.py error when trying to build libpciaccess on SNL CEE RHEL6 machine

### ISSUE THREAD

**Issue by bartlettroscoe** [spack/spack#13189]
**snl-atdm-issue**

On a Sandia CEE RHEL6 machine, I get the following build error for libpciaccess:

```
$ ./atdm-spack/spack/bin/spack install libpciaccess@0.13.5%gcc@7.2.0 cppflags="-fPIC" cxxflags="-fPIC" fflags="-fPIC"

==> libtool@2.4.6 : externally installed in /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/linux-rhel6-x86_64/gcc-7.2.0/libtool-2.4.6-njk2pzudiprn45ehlnzckluqlojzvv2j
==> libtool@2.4.6 : already registered in DB
==> pkgconf is already installed in /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/linux-rhel6-x86_64/gcc-7.2.0/pkgconf-1.5.4-uf5tave4hl3qvjyoxgzarmymvn5smof7
==> util-macros@1.19.1 : externally installed in /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/linux-rhel6-x86_64/gcc-7.2.0/util-macros-1.19.1-rjakshkke76bd6gyseig6y5shdrnv3kn
==> util-macros@1.19.1 : already registered in DB
==> Installing libpciaccess
==> Searching for binary cache of libpciaccess
==> Finding buildcaches in /scratch/rabartl/Spack.base2/atdm-spack/atdm-spack-tarball-mirror/build_cache
==> No binary for libpciaccess found: installing from source
==> Error: AssertionError: 

/scratch/rabartl/Spack.base2/atdm-spack/spack/lib/spack/spack/build_environment.py:784, in child_process:
        781            tb_string = traceback.format_exc()
        782
        783            # build up some context from the offending package so we can
  >>    784            # show that, too.
        785            package_context = get_package_context(tb)
        786
        787            build_log = None
```

I have no idea how to debug this.  This is pointing to a comment, not even a line of code.

Spack version:

```
42386db "Use Package.headers for -I options (#10623)"
Author: Massimiliano Culpo <massimiliano.culpo@gmail.com>
Date:   Tue Feb 26 19:42:30 2019 +0100 (8 months ago)

M       lib/spack/env/cc
M       lib/spack/llnl/util/filesystem.py
M       lib/spack/spack/build_environment.py
M       lib/spack/spack/test/build_environment.py
M       lib/spack/spack/test/conftest.py
M       lib/spack/spack/test/llnl/util/filesystem.py
M       var/spack/repos/builtin/packages/libxml2/package.py
```

**Comment by alalazo** [comment]
Can you try posting the same command with debug info on (`spack -d ...`)? Usually when it stops there it's because you got an exception during handling of another exception...

**Comment by bartlettroscoe** [comment]
Below is the detailed output from:

```
$ ./atdm-spack/spack/bin/spack -d install \
    libpciaccess@0.13.5%gcc@7.2.0 cppflags="-fPIC" cxxflags="-fPIC" fflags="-fPIC"
```

<details>

<summary><b>Detailed debug output:</b> (click to expand)</summary>

```
==> [2019-10-11-10:05:24.956476] Reading config file /scratch/rabartl/Spack.base2/atdm-spack/spack/etc/spack/defaults/modules.yaml
==> [2019-10-11-10:05:24.965981] Reading config file /scratch/rabartl/Spack.base2/atdm-spack/spack/etc/spack/defaults/linux/modules.yaml
==> [2019-10-11-10:05:24.968959] Reading config file /scratch/rabartl/Spack.base2/atdm-spack/spack/etc/spack/modules.yaml
==> [2019-10-11-10:05:24.970304] Reading config file /scratch/rabartl/Spack.base2/common_options.yaml
==> [2019-10-11-10:05:24.972052] Reading config file /scratch/rabartl/Spack.base2/packages_tools_for_mpi_build.yaml
==> [2019-10-11-10:05:24.989407] Reading config file /scratch/rabartl/Spack.base2/spack.yaml
==> [2019-10-11-10:05:24.999092] Reading config file /scratch/rabartl/Spack.base2/atdm-spack/spack/etc/spack/defaults/config.yaml
==> [2019-10-11-10:05:25.030974] Reading config file /scratch/rabartl/Spack.base2/atdm-spack/spack/etc/spack/defaults/repos.yaml
==> [2019-10-11-10:05:25.102070] Reading config file /scratch/rabartl/Spack.base2/atdm-spack/spack/etc/spack/defaults/packages.yaml
==> [2019-10-11-10:05:25.119190] READ LOCK: /scratch/rabartl/Spack.base2/.spack/cache/tags/.builtin-index.json.lock[0:0] [Acquiring]
==> [2019-10-11-10:05:25.119326] READ LOCK: /scratch/rabartl/Spack.base2/.spack/cache/tags/.builtin-index.json.lock[0:0] [Acquired]
==> [2019-10-11-10:05:25.120754] READ LOCK: /scratch/rabartl/Spack.base2/.spack/cache/tags/.builtin-index.json.lock[0:0] [Released]
==> [2019-10-11-10:05:25.123097] READ LOCK: /scratch/rabartl/Spack.base2/.spack/cache/patches/.builtin-index.json.lock[0:0] [Acquiring]
==> [2019-10-11-10:05:25.123222] READ LOCK: /scratch/rabartl/Spack.base2/.spack/cache/patches/.builtin-index.json.lock[0:0] [Acquired]
==> [2019-10-11-10:05:25.165918] READ LOCK: /scratch/rabartl/Spack.base2/.spack/cache/patches/.builtin-index.json.lock[0:0] [Released]
==> [2019-10-11-10:05:25.168350] READ LOCK: /scratch/rabartl/Spack.base2/.spack/cache/providers/.builtin-index.json.lock[0:0] [Acquiring]
==> [2019-10-11-10:05:25.168475] READ LOCK: /scratch/rabartl/Spack.base2/.spack/cache/providers/.builtin-index.json.lock[0:0] [Acquired]
==> [2019-10-11-10:05:25.207428] READ LOCK: /scratch/rabartl/Spack.base2/.spack/cache/providers/.builtin-index.json.lock[0:0] [Released]
==> [2019-10-11-10:05:25.244752] Reading config file /scratch/rabartl/Spack.base2/atdm-spack/spack/etc/spack/compilers.yaml
==> [2019-10-11-10:05:25.299197] DATABASE LOCK TIMEOUT: 120s
==> [2019-10-11-10:05:25.299268] PACKAGE LOCK TIMEOUT: No timeout
==> [2019-10-11-10:05:25.300664] WRITE LOCK: /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/.spack-db/prefix_lock[1446588495073676357:1] [Acquiring]
==> [2019-10-11-10:05:25.300795] WRITE LOCK: /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/.spack-db/prefix_lock[1446588495073676357:1] [Acquired]
==> [2019-10-11-10:05:25.300878] READ LOCK: /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/.spack-db/lock[0:0] [Acquiring]
==> [2019-10-11-10:05:25.300999] READ LOCK: /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/.spack-db/lock[0:0] [Acquired]
==> [2019-10-11-10:05:25.345616] READ LOCK: /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/.spack-db/lock[0:0] [Released]
==> [2019-10-11-10:05:25.346655] WRITE LOCK: /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/.spack-db/prefix_lock[1446588495073676357:1] [Released]
==> [2019-10-11-10:05:25.360795] link /scratch/rabartl/Spack.base2/atdm-spack/spack/var/spack/stage/libpciaccess-0.13.5-fatkcjgscemiv5xuparjkrkwpu2owadt -> /tmp/rabartl/spack-stage/spack-stage-6Z6Jmi
==> [2019-10-11-10:05:25.361292] READ LOCK: /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/.spack-db/prefix_lock[1446588495073676357:1] [Acquiring]
==> [2019-10-11-10:05:25.361415] READ LOCK: /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/.spack-db/prefix_lock[1446588495073676357:1] [Acquired]
==> [2019-10-11-10:05:25.361874] READ LOCK: /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/.spack-db/prefix_lock[1446588495073676357:1] [Released]
==> [2019-10-11-10:05:25.361946] Installing libpciaccess dependencies
==> [2019-10-11-10:05:25.368390] WRITE LOCK: /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/.spack-db/prefix_lock[5817973601099152261:1] [Acquiring]
==> [2019-10-11-10:05:25.368520] WRITE LOCK: /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/.spack-db/prefix_lock[5817973601099152261:1] [Acquired]
==> [2019-10-11-10:05:25.368592] WRITE LOCK: /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/.spack-db/prefix_lock[5817973601099152261:1] [Released]
==> [2019-10-11-10:05:25.373108] link /scratch/rabartl/Spack.base2/atdm-spack/spack/var/spack/stage/pkgconf-1.5.4-uf5tave4hl3qvjyoxgzarmymvn5smof7 -> /tmp/rabartl/spack-stage/spack-stage-8gCdwE
==> [2019-10-11-10:05:25.380055] READ LOCK: /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/.spack-db/prefix_lock[5817973601099152261:1] [Acquiring]
==> [2019-10-11-10:05:25.380196] READ LOCK: /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/.spack-db/prefix_lock[5817973601099152261:1] [Acquired]
==> [2019-10-11-10:05:25.386581] READ LOCK: /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/.spack-db/prefix_lock[5817973601099152261:1] [Released]
==> [2019-10-11-10:05:25.362051] libtool@2.4.6 : externally installed in /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/linux-rhel6-x86_64/gcc-7.2.0/libtool-2.4.6-njk2pzudiprn45ehlnzckluqlojzvv2j
==> [2019-10-11-10:05:25.365062] libtool@2.4.6 : already registered in DB
==> [2019-10-11-10:05:25.386295] pkgconf is already installed in /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/linux-rhel6-x86_64/gcc-7.2.0/pkgconf-1.5.4-uf5tave4hl3qvjyoxgzarmymvn5smof7
==> [2019-10-11-10:05:25.386694] util-macros@1.19.1 : externally installed in /scratch/rabartl/Spack.base2/atdm-spack/spack/opt/spack/linux-rhel6-x86_64/gcc-7.2.0/util-macros-1.19.1-rjakshkke76bd6gyseig6y5shdrnv3kn
==> [2019-10-11-10:05:25.390103] util-macros@1.19.1 : already registered in DB
==> [2019-10-11-10:05:25.390177] Installing libpciaccess
==> [2019-10-11-10:05:25.390216] Searching for binary cache of libpciaccess
==> [2019-10-11-10:05:25.390514] Finding buildcaches in /scratch/rabartl/Spack.base2/atdm-spack/atdm-spack-tarball-mirror/build_cache
==> [2019-10-11-10:05:25.391042] No binary for libpciaccess found: installing from source
==> [2019-10-11-10:05:25.464300] Error: AssertionError: 

/scratch/rabartl/Spack.base2/atdm-spack/spack/lib/spack/spack/build_environment.py:784, in child_process:
        781            tb_string = traceback.format_exc()
        782
        783            # build up some context from the offending package so we can
  >>    784            # show that, too.
        785            package_context = get_package_context(tb)
        786
        787            build_log = None


Traceback (most recent call last):
  File "/scratch/rabartl/Spack.base2/atdm-spack/spack/lib/spack/spack/build_environment.py", line 766, in child_process
    setup_package(pkg, dirty=dirty)
  File "/scratch/rabartl/Spack.base2/atdm-spack/spack/lib/spack/spack/build_environment.py", line 675, in setup_package
    set_build_environment_variables(pkg, spack_env, dirty)
  File "/scratch/rabartl/Spack.base2/atdm-spack/spack/lib/spack/spack/build_environment.py", line 308, in set_build_environment_variables
    build_prefixes      = [dep.prefix for dep in build_deps]
  File "/scratch/rabartl/Spack.base2/atdm-spack/spack/lib/spack/spack/spec.py", line 1270, in prefix
    self.prefix = spack.store.layout.path_for_spec(self)
  File "/scratch/rabartl/Spack.base2/atdm-spack/spack/lib/spack/spack/directory_layout.py", line 78, in path_for_spec
    assert(not path.startswith(self.root))
AssertionError

```

</details>
 
FYI: This build used to work just fine on this same SNL CEE RHEL6 machine a few months ago.  Therefore, something must have changed on the system but I don't understand why this should be so fragile since everything else we are building there (e.g. Trilinos using a non-Spack stack of tools) is still building just fine.  There just seems to be some significant facility to many of these package installs.

**Comment by bartlettroscoe** [comment]
I can verify that this was resolved when we upgraded to spack v0.13.0 (and build with `arch=linux-rhel6-x86_64`).

Closing as complete.
